### PR TITLE
Document pack_hook reference-cycle caveat for saved tensor hooks

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #189025

The saved-tensors-hooks docs show pack hooks calling `.detach()` on their
input but never explain why. Add a note to the "Registering default hooks
for saved tensors" section spelling out the reason and its precise semantics.

With the `saved_tensors_hooks` context manager, `pack_hook` receives a live
tensor that still carries its `grad_fn` (unlike per-tensor `register_hooks`,
which receives an already-detached tensor). Retaining that tensor when the
saved tensor is a graph output forms a reference cycle: the output's `grad_fn`
owns the SavedVariable, which would then own a tensor referring back to that
same `grad_fn`. This is exactly the cycle PyTorch's internals avoid by storing
`tensor_data()` for saved outputs (see the comment in saved_variable.cpp).

The note also corrects a subtle point about reclamation: the cycle does not
leak under a normal `backward()` (retain_graph=False releases the saved tensor
and breaks the cycle), but it is unreclaimable -- even by an explicit
gc.collect() -- when the graph is built without running backward or with
retain_graph=True, because PyTorch's tensor traverse deliberately does not walk
the autograd graph (pytorch/pytorch#7343). Detaching is lossless: autograd
stashes the grad_fn separately in save_metadata and restores it on unpack, so
first- and higher-order gradients are unaffected. Offload-style hooks that
return a freshly computed tensor are unaffected because pack hooks run with
gradient tracking disabled.

Also adds a one-line pointer from the `torch.autograd` API reference.

Test Plan: docs-only change; verified with lintrunner:

```
lintrunner -a docs/source/autograd.md docs/source/notes/autograd.md
```

Authored-with: Claude (Anthropic AI assistant)

### CODE REVIEW

**Review by albanD — APPROVED** [review]
Thanks!

**Review by soulitzer — APPROVED** [review]
